flink-staging/flink-streaming/flink-streaming-core/src/main/java/org/apache/flink/streaming/runtime/io/StreamInputProcessor.java

```diff
@@ -0,0 +1,186 @@
+ *
+ * <p>
+ * This also keeps track of {@link Watermark} events and forwards them to event subscribers
+ * once the {@link Watermark} from all inputs advances.
+ * 
+ * @param <IN> The type of the record that can be read with this record reader.
+ */
+public class StreamInputProcessor<IN> extends AbstractReader implements ReaderBase, StreamingReader {
+
+	@SuppressWarnings("unused")
+	private static final Logger LOG = LoggerFactory.getLogger(StreamInputProcessor.class);
+
+	private final RecordDeserializer<DeserializationDelegate>[] recordDeserializers;
+
+	private RecordDeserializer<DeserializationDelegate> currentRecordDeserializer;
+
+	// We need to keep track of the channel from which a buffer came, so that we can
+	// appropriately map the watermarks to input channels
+	int currentChannel = -1;
+
+	private boolean isFinished;
+
+	private final BarrierBuffer barrierBuffer;
+
+	private long[] watermarks;
+	private long lastEmittedWatermark;
+
+	private DeserializationDelegate deserializationDelegate;
+
+	@SuppressWarnings("unchecked")
+	public StreamInputProcessor(InputGate[] inputGates, TypeSerializer<IN> inputSerializer) {
+		super(InputGateUtil.createInputGate(inputGates));
+
+		barrierBuffer = new BarrierBuffer(inputGate, this);
+
+		StreamRecordSerializer<IN> inputRecordSerializer = new StreamRecordSerializer<IN>(inputSerializer);
+		this.deserializationDelegate = new NonReusingDeserializationDelegate(inputRecordSerializer);
+
+		// Initialize one deserializer per input channel
+		this.recordDeserializers = new SpillingAdaptiveSpanningRecordDeserializer[inputGate.getNumberOfInputChannels()];
+		for (int i = 0; i < recordDeserializers.length; i++) {
+			recordDeserializers[i] = new SpillingAdaptiveSpanningRecordDeserializer<DeserializationDelegate>();
+		}
+
+		watermarks = new long[inputGate.getNumberOfInputChannels()];
+		for (int i = 0; i < inputGate.getNumberOfInputChannels(); i++) {
+			watermarks[i] = Long.MIN_VALUE;
+		}
+		lastEmittedWatermark = Long.MIN_VALUE;
+	}
+
+	@SuppressWarnings("unchecked")
+	public boolean processInput(OneInputStreamOperator<IN, ?> streamOperator) throws Exception {
+		if (isFinished) {
+			return false;
+		}
+
+		while (true) {
+			if (currentRecordDeserializer != null) {
+				DeserializationResult result = currentRecordDeserializer.getNextRecord(deserializationDelegate);
+
+				if (result.isBufferConsumed()) {
+					currentRecordDeserializer.getCurrentBuffer().recycle();
+					currentRecordDeserializer = null;
+				}
+
+				if (result.isFullRecord()) {
+					Object recordOrWatermark = deserializationDelegate.getInstance();
+
+					if (recordOrWatermark instanceof Watermark) {
+						Watermark mark = (Watermark) recordOrWatermark;
+						long watermarkMillis = mark.getTimestamp();
+						if (watermarkMillis > watermarks[currentChannel]) {
+							watermarks[currentChannel] = watermarkMillis;
+							long newMinWatermark = Long.MAX_VALUE;
+							for (long watermark : watermarks) {
```

Comment:
The whole thing with watermarks does not really work with iterations, correct. We'll have to figure out a way to do it.